server/app/jobs/grid_service_scheduler_worker.rb

```diff
@@ -9,58 +9,59 @@ def initialize(autostart = true)
 
   def watch
     loop do
-      self.check_deploy_queue
+      if service_deploy = self.check_deploy_queue
+        self.deploy(service_deploy)
+      end
       sleep 1
     end
   end
 
+  # Fetch deploy from queue, and return it if it should be run
+  #
+  # @return [GridServiceDeploy, nil] deploy to run
   def check_deploy_queue
     service_deploy = fetch_deploy_item
-    return unless service_deploy
+    return nil unless service_deploy
 
     with_dlock("check_deploy_queue:#{service_deploy.grid_service_id}", 10) do
-      service_deploy.reload
-      if service_deploy.grid_service.running? || service_deploy.grid_service.initialized?
-        self.perform(service_deploy)
+      if service_deploy.grid_service.deploy_running?
+        info "delaying #{service_deploy.grid_service.to_path} deploy because there is another deploy in progress"
+        return nil
+
+      elsif service_deploy.grid_service.running? || service_deploy.grid_service.initialized?
+        info "starting #{service_deploy.grid_service.to_path} deploy"
+        service_deploy.set(started_at: Time.now.utc)
+        return service_deploy
+
       else
-        service_deploy.destroy
+        info "aborting #{service_deploy.grid_service.to_path} deploy of non-running service"
+        service_deploy.abort! "service is not running"
+        return nil
       end
     end
   end
 
+  # Pick up the oldest non-queued deploy, mark it as queued, and return for processing.
+  # The caller has 30s to process the returned deploy, or it can be picked up again.
+  #
   # @return [GridServiceDeploy, NilClass]
   def fetch_deploy_item
-    GridServiceDeploy.where(started_at: nil)
+    GridServiceDeploy.any_of({:queued_at => nil}, {:queued_at.lt => 30.seconds.ago, :started_at => nil})
       .asc(:created_at)
-      .find_and_modify({:$set => {started_at: Time.now.utc}}, {new: true})
+      .find_and_modify({:$set => {:queued_at => Time.now.utc}}, {new: true})
   rescue Moped::Errors::OperationFailure
     nil
   end
 
-  def perform(service_deploy)
-    unless service_deploy.grid_service.deploying?(ignore: service_deploy.id)
-      deploy(service_deploy)
-    else
-      info "delaying #{service_deploy.grid_service.to_path} deploy because there is another deploy in progress"
-      after(30) {
-        service_deploy.set(:started_at => nil)
-      }
-    end
-  end
-
   def deploy(service_deploy)
     self.deployer(service_deploy).deploy
     self.deploy_dependant_services(service_deploy.grid_service)
   end
 
   def deploy_dependant_services(grid_service)
-    grid_service.dependant_services.each do |serv|
-      info "deploying dependent service #{serv.to_path} of deployed service #{grid_service.to_path}"
-      service_deploy = GridServiceDeploy.create(
-        grid_service: serv,
-        started_at: Time.now.utc
-      )
-      self.class.new.perform(service_deploy)
```

Comment:
This was weird... it bypassed the `check_deploy_queue` dlock, and there is nothing preventing this from running a deploy for a dependent service while a different deploy was queued/running for the same service?

I don't see why this can't just be run the normal way, via the worker queueing.